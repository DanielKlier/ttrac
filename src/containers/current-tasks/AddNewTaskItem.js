import * as React from 'react';
import {Button, Clearfix, Glyphicon} from 'react-bootstrap';
import PropTypes from 'prop-types';

class AddNewTaskItem extends React.PureComponent {

    // noinspection JSUnusedGlobalSymbols
    static propTypes = {
        onClick: PropTypes.func
    };

    render() {
        return (
            <li className="list-group-item">
                <Clearfix>
                    <span className="add-new-task-text">Add new task</span>
                    <Button className="pull-right" bsStyle="primary"
                            onClick={() => this.props.onClick()}>
                        <Glyphicon glyph="plus"/>
                    </Button>
                </Clearfix>
            </li>
        )
    }
}

export default AddNewTaskItem;
