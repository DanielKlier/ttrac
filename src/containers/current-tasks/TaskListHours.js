import * as React from 'react';
import {Label} from 'react-bootstrap';

export default props => (
    <span className="task-list-hours">
      <Label bsStyle="info">{props.hours}</Label>
    </span>
)