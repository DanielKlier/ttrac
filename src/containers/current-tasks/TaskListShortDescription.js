import * as React from 'react';

export default props => (
    <h4 className="task-list-description">
      {props.description}
    </h4>
)