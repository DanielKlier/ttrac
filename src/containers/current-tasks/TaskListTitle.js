import * as React from 'react';
import PropTypes from 'prop-types';
import ToggleDisplay from '../../components/ToggleDisplay';

export default class TaskListTitle extends React.Component {
    // noinspection JSUnusedGlobalSymbols
    static propTypes = {
        title        : PropTypes.string,
        onTextChanged: PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            isEditing: !props.title,
            text     : this.props.title || ''
        };

        this.titleInput = null;
    }

    componentWillReceiveProps(props) {

    }

    render() {

        return (
            <h4 className="task-list-description">
                <ToggleDisplay show={!this.state.isEditing && this.state.text}>
                    <span onClick={() => this.showEditInput()}>
                        {this.state.text}
                    </span>
                </ToggleDisplay>
                <ToggleDisplay show={this.state.isEditing || !this.state.text}>
                    <input type="text"
                           ref={el => {this.titleInput = el;}}
                           value={this.state.text}
                           onBlur={(e) => this.stopEditing(e.target.value)}
                           placeholder="Click to add a title"
                           onChange={
                               (e) => this.textInputChanged(e.target.value)
                           }
                    />
                </ToggleDisplay>
            </h4>
        );
    }

    showEditInput() {
        const input = this.titleInput;
        this.setState({isEditing: true}, () => {
            input.focus();
            input.select();
        });
    }

    stopEditing(value) {
        this.setState({isEditing: false});
        this.props.onTextChanged(value);
    }

    textInputChanged(value) {
        this.setState({text: value});
    }
}