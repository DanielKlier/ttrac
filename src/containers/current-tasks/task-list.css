@keyframes pulsating-border {
    from {
        box-shadow: 0 0 5px rgba(255, 0, 2, 1) inset;
    }
    to {
        box-shadow: 0 0 0 rgba(255, 0, 2, 0) inset;
    }
}

.task-list-item {
    display: flex;
}

.task-list-item.running {
    animation: 1s ease infinite alternate pulsating-border;
}

.task-list-hours {
    font-size: 24px;
}

.task-list-issue {
    display: block;
}

.task-list-item-main {
    width: 100%;
    margin-right: 10px;
    overflow-x: hidden;
}

.task-list-description {
    margin-bottom: 5px;
    white-space: nowrap;
    text-overflow: ellipsis;
    width: 100%;
    overflow: hidden;
}

.task-list-item-badges {
    margin-right: 10px;
    flex-shrink: 0;
}

.task-list-item-buttons {
    flex-shrink: 0;
}

.add-new-task-text {
    line-height: 34px;
}
