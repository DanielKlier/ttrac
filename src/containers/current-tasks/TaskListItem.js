import * as React from 'react';
import PropTypes from 'prop-types';
import TaskListHours from './TaskListHours';
import TaskListIssue from './TaskListIssue';
import {Button, Glyphicon} from 'react-bootstrap';
import TaskListTitle from './TaskListTitle';
import TimeCounter from './TimeCounter';

class TaskListItem extends React.PureComponent {
    // noinspection JSUnusedGlobalSymbols
    static PropTypes = {
        elapsedTime           : PropTypes.number,
        jiraIssue             : PropTypes.string,
        title                 : PropTypes.string,
        taskIsRunning         : PropTypes.bool,
        runningTask           : PropTypes.object,
        onTaskTitleChanged    : PropTypes.func.isRequired,
        onStartProgressClicked: PropTypes.func.isRequired,
        onStopProgressClicked : PropTypes.func.isRequired,
        onDeleteTaskClicked   : PropTypes.func.isRequired
    };

    render() {
        const props      = this.props;
        const classNames = [
            'list-group-item',
            'task-list-item'
        ];
        if (props.taskIsRunning) {
            classNames.push('running');
        }
        const className = classNames.join(' ');

        return (
            <li className={className}>
                <div className="task-list-item-badges">
                    {props.taskIsRunning &&
                    <TimeCounter startDate={props.runningTask.startDate}/>
                    }
                    {!props.taskIsRunning &&
                    <TaskListHours elapsedTime={props.elapsedTime}
                                   className="pull-left"/>
                    }
                    {props.jiraIssue &&
                    <TaskListIssue jiraIssue={props.jiraIssue}/>}
                </div>
                <div className="task-list-item-main">
                    <TaskListTitle title={props.title}
                                   onTextChanged={
                                       text => props.onTaskTitleChanged(text)
                                   }/>
                </div>
                <div className="task-list-item-buttons">
                    <Button onClick={() => props.onDeleteTaskClicked()}>
                        <Glyphicon glyph="trash"/>
                    </Button>
                    {props.taskIsRunning === false &&
                    <Button onClick={() => props.onStartProgressClicked()}>
                        <Glyphicon glyph="time"/>
                    </Button>
                    }
                    {props.taskIsRunning === true &&
                    <Button onClick={() => props.onStopProgressClicked()}>
                        <Glyphicon glyph="pause"/>
                    </Button>
                    }
                </div>
            </li>
        );
    }
}

export default TaskListItem;
