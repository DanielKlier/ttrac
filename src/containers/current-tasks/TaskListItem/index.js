import * as React from 'react';
import PropTypes from 'prop-types';
import {identity} from 'lodash';
import TaskListHours from './TaskListHours';
import TaskListIssue from './TaskListIssue';
import TaskListTitle from './TaskListTitle';
import TimeCounter from './TimeCounter';
import StartProgressButton from './StartProgressButton';
import StopProgressButton from './StopProgressButton';
import DeleteButton from '../../../components/DeleteButton';

class TaskListItem extends React.PureComponent {
    // noinspection JSUnusedGlobalSymbols
    static propTypes = {
        elapsedTime: PropTypes.number,
        jiraIssue: PropTypes.string,
        title: PropTypes.string,
        taskIsRunning: PropTypes.bool,
        runningTask: PropTypes.object,
        onTaskTitleChanged: PropTypes.func,
        onStartProgressClicked: PropTypes.func,
        onStopProgressClicked: PropTypes.func,
        onDeleteTaskClicked: PropTypes.func
    };

    // noinspection JSUnusedGlobalSymbols
    static defaultProps = {
        elapsedTime: 0,
        jiraIssue: '',
        title: '',
        taskIsRunning: false,
        runningTask: null,
        onTaskTitleChanged: identity,
        onStartProgressClicked: identity,
        onStopProgressClicked: identity,
        onDeleteTaskClicked: identity
    };

    render() {
        const props      = this.props;
        const classNames = [
            'list-group-item',
            'task-list-item'
        ];
        if (props.taskIsRunning) {
            classNames.push('running');
        }
        const className = classNames.join(' ');

        return (
            <li className={className}>
                <div className="task-list-item-badges">
                    {props.taskIsRunning &&
                    <TimeCounter startDate={props.runningTask.startDate}/>
                    }
                    {!props.taskIsRunning &&
                    <TaskListHours elapsedTime={props.elapsedTime}
                                   className="pull-left"/>
                    }
                    {props.jiraIssue &&
                    <TaskListIssue jiraIssue={props.jiraIssue}/>}
                </div>
                <div className="task-list-item-main">
                    <TaskListTitle title={props.title}
                                   onTextChanged={
                                       text => props.onTaskTitleChanged(text)
                                   }/>
                </div>
                <div className="task-list-item-buttons">
                    <DeleteButton onClick={() => props.onDeleteTaskClicked()}/>
                    {props.taskIsRunning === false &&
                    <StartProgressButton onClick={() => props.onStartProgressClicked()}/>
                    }
                    {props.taskIsRunning === true &&
                    <StopProgressButton onClick={() => props.onStopProgressClicked()}/>
                    }
                </div>
            </li>
        );
    }
}

export default TaskListItem;
