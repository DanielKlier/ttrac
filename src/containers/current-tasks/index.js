import React from 'react';
import {connect} from 'react-redux';
import {ListGroup} from 'react-bootstrap';
import TaskListItem from './TaskListItem';
import './task-list.css';
import AddNewTaskItem from './AddNewTaskItem';
import {
    createNewTask,
    deleteTask,
    setTaskTitle,
    startTaskProgress,
    stopTaskProgress
} from '../../actions/';
import getCurrentTasks from '../../selectors/tasks/getCurrentTasks';

const CurrentTasks = props => (
    <div>
        <h1>Time Tracking</h1>
        <ListGroup componentClass="ul">
            {
                props.currentTasks.map(t => (
                    <TaskListItem {...t}
                                  key={t.id}
                                  taskIsRunning={
                                      !!props.runningTask &&
                                      t.id === props.runningTask.taskId
                                  }
                                  runningTask={props.runningTask}
                                  onTaskTitleChanged={title =>
                                      props.onTaskTitleChanged(t.id, title)
                                  }
                                  onStartProgressClicked={() =>
                                      props.onStartProgressClicked(t.id)
                                  }
                                  onStopProgressClicked={() =>
                                      props.runningTask &&
                                      props.onStopProgressClicked(
                                          t.id, props.runningTask.startDate
                                      )
                                  }
                                  onDeleteTaskClicked={() =>
                                      props.onDeleteTaskClicked(t.id)
                                  }
                    />
                ))
            }
            <AddNewTaskItem
                onClick={() => props.onAddNewTaskClick()}/>
        </ListGroup>
    </div>
);

const mapStateToProps = (state) => ({
    currentTasks: getCurrentTasks(state).map(task => {

        const elapsedTime = (task.timeLogIds || [])
            .map(logId => state.app.timeLogs.byId[logId])
            .reduce((time, log) => time + log.stopDate - log.startDate, 0);

        return {
            ...task,
            elapsedTime
        };
    }),
    runningTask : state.app.runningTask
});

const mapDispatchToProps = dispatch => ({
    onAddNewTaskClick     : () => dispatch(createNewTask()),
    onTaskTitleChanged    : (taskId, text) => dispatch(
        setTaskTitle(taskId, text)),
    onStartProgressClicked: (taskId) => dispatch(startTaskProgress(taskId, Date.now())),
    onStopProgressClicked : (taskId, startDate) => dispatch(
        stopTaskProgress(taskId, startDate, Date.now())
    ),
    onDeleteTaskClicked   : (taskId) => dispatch(deleteTask(taskId))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CurrentTasks);
