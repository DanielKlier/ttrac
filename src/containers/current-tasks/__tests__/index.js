import React from 'react';
import {CurrentTasks, mapDispatchToProps, mapStateToProps} from '../../current-tasks';
import {mount} from 'enzyme';
import toJSON from 'enzyme-to-json';

test('it behaves correctly', () => {

    const handleAddNewTask      = jest.fn();
    const handleTaskTitleChange = jest.fn();
    const handleTaskStopped     = jest.fn();
    const handleTaskStarted     = jest.fn();
    const handleDeleteTask      = jest.fn();

    const tasks = [
        {id: '1', elapsedTime: 2000},
        {id: '2'}
    ];

    const runningTask = {taskId: '1', startDate: Date.parse('2017-03-09 18:00:00')};

    const component = mount(
        <CurrentTasks
            currentTasks={tasks}
            runningTask={runningTask}
            onAddNewTaskClick={handleAddNewTask}
            onTaskTitleChanged={handleTaskTitleChange}
            onStartProgressClicked={handleTaskStarted}
            onStopProgressClicked={handleTaskStopped}
            onDeleteTaskClicked={handleDeleteTask}
        />
    );

    expect(toJSON(component)).toMatchSnapshot();

    component.find('TaskListItem StopProgressButton').simulate('click');
    expect(handleTaskStopped).toHaveBeenCalled();

    component.find('AddNewTaskItem Button').simulate('click');
    expect(handleAddNewTask).toHaveBeenCalled();

    component.find('StartProgressButton').at(0).simulate('click');
    expect(handleTaskStarted).toHaveBeenCalled();

    component.find('DeleteButton').at(0).simulate('click');
    expect(handleDeleteTask).toHaveBeenCalled();

    component.find('TaskListTitle')
        .at(0)
        .find('.title-input')
        .simulate('change', {target: {value: 'hello'}});

    component.find('TaskListTitle').at(0).find('.title-input').simulate('blur');

    expect(handleTaskTitleChange).toHaveBeenCalled();
});

test('mapStateToProps', () => {
    const state = {
        app: {
            runningTask: {taskId: 1, startDate: 0},
            tasks: {
                byId: {
                    1: {id: 1},
                    2: {id: 2, timeLogIds: [1]}
                },
                allIds: [1, 2],
                deletedIds: []
            },
            timeLogs: {
                byId: {
                    1: {id: 1, stopDate: 2000, startDate: 0}
                }
            }
        }
    };

    expect(mapStateToProps(state)).toEqual({
        currentTasks: [
            {
                id: 1, elapsedTime: 0
            },
            {
                id: 2, elapsedTime: 2000, timeLogIds: [1]
            }
        ],
        runningTask: {taskId: 1, startDate: 0}
    })
});

test('mapDispatchToProps', () => {

    const spy = jest.fn();

    const {onAddNewTaskClick} = mapDispatchToProps(spy);
    onAddNewTaskClick();
    expect(spy).toHaveBeenCalled();
    spy.mockReset();

    const {onTaskTitleChanged} = mapDispatchToProps(spy);
    onTaskTitleChanged();
    expect(spy).toHaveBeenCalled();
    spy.mockReset();

    const {onStartProgressClicked} = mapDispatchToProps(spy);
    onStartProgressClicked();
    expect(spy).toHaveBeenCalled();
    spy.mockReset();

    const {onStopProgressClicked} = mapDispatchToProps(spy);
    onStopProgressClicked();
    expect(spy).toHaveBeenCalled();
    spy.mockReset();

    const {onDeleteTaskClicked} = mapDispatchToProps(spy);
    onDeleteTaskClicked();
    expect(spy).toHaveBeenCalled();
    spy.mockReset();
});
