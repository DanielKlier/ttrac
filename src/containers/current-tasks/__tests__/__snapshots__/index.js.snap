// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it behaves correctly 1`] = `
<Component
  currentTasks={
    Array [
      Object {
        "elapsedTime": 2000,
        "id": "1",
      },
      Object {
        "id": "2",
      },
    ]
  }
  onAddNewTaskClick={[Function]}
  onDeleteTaskClicked={[Function]}
  onStartProgressClicked={[Function]}
  onStopProgressClicked={[Function]}
  onTaskTitleChanged={[Function]}
  runningTask={
    Object {
      "startDate": 1489078800000,
      "taskId": "1",
    }
  }
>
  <div>
    <h1>
      Time Tracking
    </h1>
    <ListGroup
      bsClass="list-group"
      componentClass="ul"
    >
      <ul
        className="list-group"
      >
        <TaskListItem
          elapsedTime={2000}
          id="1"
          key="1"
          onDeleteTaskClicked={[Function]}
          onStartProgressClicked={[Function]}
          onStopProgressClicked={[Function]}
          onTaskTitleChanged={[Function]}
          runningTask={
            Object {
              "startDate": 1489078800000,
              "taskId": "1",
            }
          }
          taskIsRunning={true}
        >
          <li
            className="list-group-item task-list-item running"
          >
            <div
              className="task-list-item-badges"
            >
              <TimeCounter
                startDate={1489078800000}
              >
                <span
                  className="task-list-hours"
                >
                  <Label
                    bsClass="label"
                    bsStyle="warning"
                  >
                    <span
                      className="label label-warning"
                    >
                      0s
                    </span>
                  </Label>
                </span>
              </TimeCounter>
            </div>
            <div
              className="task-list-item-main"
            >
              <TaskListTitle
                onTextChanged={[Function]}
                title={undefined}
              >
                <h4
                  className="task-list-description"
                >
                  <toggleDisplay
                    show={false}
                  >
                    <span
                      className="hidden"
                    >
                      <span
                        className="title"
                        onClick={[Function]}
                      />
                    </span>
                  </toggleDisplay>
                  <toggleDisplay
                    show={true}
                  >
                    <span
                      className=""
                    >
                      <input
                        className="title-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="Click to add a title"
                        type="text"
                        value=""
                      />
                    </span>
                  </toggleDisplay>
                </h4>
              </TaskListTitle>
            </div>
            <div
              className="task-list-item-buttons"
            >
              <DeleteButton
                onClick={[Function]}
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Glyphicon
                      bsClass="glyphicon"
                      glyph="trash"
                    >
                      <span
                        className="glyphicon glyphicon-trash"
                      />
                    </Glyphicon>
                  </button>
                </Button>
              </DeleteButton>
              <StopProgressButton
                onClick={[Function]}
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Glyphicon
                      bsClass="glyphicon"
                      glyph="pause"
                    >
                      <span
                        className="glyphicon glyphicon-pause"
                      />
                    </Glyphicon>
                  </button>
                </Button>
              </StopProgressButton>
            </div>
          </li>
        </TaskListItem>
        <TaskListItem
          id="2"
          key="2"
          onDeleteTaskClicked={[Function]}
          onStartProgressClicked={[Function]}
          onStopProgressClicked={[Function]}
          onTaskTitleChanged={[Function]}
          runningTask={
            Object {
              "startDate": 1489078800000,
              "taskId": "1",
            }
          }
          taskIsRunning={false}
        >
          <li
            className="list-group-item task-list-item"
          >
            <div
              className="task-list-item-badges"
            >
              <TaskListHours
                className="pull-left"
                elapsedTime={undefined}
              >
                <span
                  className="task-list-hours"
                >
                  <Label
                    bsClass="label"
                    bsStyle="info"
                  >
                    <span
                      className="label label-info"
                    >
                      0s
                    </span>
                  </Label>
                </span>
              </TaskListHours>
            </div>
            <div
              className="task-list-item-main"
            >
              <TaskListTitle
                onTextChanged={[Function]}
                title={undefined}
              >
                <h4
                  className="task-list-description"
                >
                  <toggleDisplay
                    show={false}
                  >
                    <span
                      className="hidden"
                    >
                      <span
                        className="title"
                        onClick={[Function]}
                      />
                    </span>
                  </toggleDisplay>
                  <toggleDisplay
                    show={true}
                  >
                    <span
                      className=""
                    >
                      <input
                        className="title-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="Click to add a title"
                        type="text"
                        value=""
                      />
                    </span>
                  </toggleDisplay>
                </h4>
              </TaskListTitle>
            </div>
            <div
              className="task-list-item-buttons"
            >
              <DeleteButton
                onClick={[Function]}
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Glyphicon
                      bsClass="glyphicon"
                      glyph="trash"
                    >
                      <span
                        className="glyphicon glyphicon-trash"
                      />
                    </Glyphicon>
                  </button>
                </Button>
              </DeleteButton>
              <StartProgressButton
                onClick={[Function]}
              >
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-default"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Glyphicon
                      bsClass="glyphicon"
                      glyph="time"
                    >
                      <span
                        className="glyphicon glyphicon-time"
                      />
                    </Glyphicon>
                  </button>
                </Button>
              </StartProgressButton>
            </div>
          </li>
        </TaskListItem>
        <AddNewTaskItem
          onClick={[Function]}
        >
          <li
            className="list-group-item"
          >
            <Clearfix
              bsClass="clearfix"
              componentClass="div"
            >
              <div
                className="clearfix"
              >
                <span
                  className="add-new-task-text"
                >
                  Add new task
                </span>
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="primary"
                  className="pull-right"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="pull-right btn btn-primary"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <Glyphicon
                      bsClass="glyphicon"
                      glyph="plus"
                    >
                      <span
                        className="glyphicon glyphicon-plus"
                      />
                    </Glyphicon>
                  </button>
                </Button>
              </div>
            </Clearfix>
          </li>
        </AddNewTaskItem>
      </ul>
    </ListGroup>
  </div>
</Component>
`;
